File: examples/java-junit-gradle/src/test/java/examples/HelloVarSteps.java
package examples;

import dev.varar.Registrar;
import dev.varar.State;
import dev.varar.StateBinder;
import dev.varar.StepDefinitions;

public final class HelloVarSteps implements StepDefinitions {

    record Ctx(String greeting, int result) implements State {}

    @Override
    public void defineSteps(Registrar registrar) {
        StateBinder<Ctx> s = registrar.steps(() -> new Ctx("", 0));

        s.stimulus("I greet {string}", (Ctx ctx, String name) -> new Ctx("Hello, " + name + "!", ctx.result()));
        s.sensor("the greeting should be {string}", (Ctx ctx, String expected) -> ctx.greeting());
        s.stimulus("expression `{int}+{int}`", (Ctx ctx, Integer a, Integer b) -> new Ctx(ctx.greeting(), a + b));
        s.sensor("evaluate to `{int}`", (Ctx ctx, Integer expected) -> ctx.result());
    }
}

File: java/examples-kotlin-kotest/src/test/kotlin/examples/hello-var.steps.kt
@file:JvmName("HelloVarSteps")

package examples

import com.oselvar.varkt.defineState
import com.oselvar.varkt.sensor
import com.oselvar.varkt.stimulus

data class HelloCtx(val greeting: String = "", val result: Int = 0)

val helloVarSteps = defineState(::HelloCtx) {
    stimulus("I greet {string}") { name: String -> copy(greeting = "Hello, $name!") }
    sensor("the greeting should be {string}") { _: String -> greeting }
    stimulus("expression `{int}+{int}`") { a: Int, b: Int -> copy(result = a + b) }
    sensor("evaluate to `{int}`") { _: Int -> result }
}
